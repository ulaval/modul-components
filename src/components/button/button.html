<button :class="[ componentName,
                { 'is-primary': type!='secondary',
                'is-secondary': type=='secondary',
                'is-selected': state=='selected',
                'is-waiting': state=='waiting',
                'is-disabled': state=='disabled' } ]"
        v-m-ripple-effect="!isDisabled"
        :disabled="isDisabled">
    <span :class="[ componentName + '__icon', 'is-left']" v-if="hasIconeLeft">
        <slot name="icon-left"></slot>
    </span>
    <span :class="[ componentName + '__text']">
        <slot>{{ componentName }}</slot>
        <span :class="[ componentName + '__more-info' ]" v-if="hasMoreInfo">
            <slot name="more-info"></slot>
        </span>
    </span>
    <span :class="[ componentName + '__icon', 'is-right']" v-if="hasIconeRight">
        <slot name="icon-right"></slot>
    </span>
</button>