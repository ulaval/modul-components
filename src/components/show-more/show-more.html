<div class="m-show-more"
     v-if="isVisible">
    <div class="m-show-more__status"
         ref="countTotal">{{ status }}</div>
    <m-progress v-if="isProgressVisible"
                :value="visiblePourcentage"
                :border-radius="false"
                ref="progress"></m-progress>
    <m-button v-if="isButtonVisible"
              :skin="buttonSkin"
              @click="showMore"
              :waiting="loading"
              ref="button">{{ buttonLabel }}</m-button>
</div>
