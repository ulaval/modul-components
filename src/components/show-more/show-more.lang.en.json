{
    "m-show-more": {
        "status": "%(nbVisible)s of %(nbTotal)s",
        "button-label": "Show more results"
    }
}
