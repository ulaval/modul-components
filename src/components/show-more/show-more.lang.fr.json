{
    "m-show-more": {
        "status": "%(nbVisible)s sur %(nbTotal)s",
        "button-label": "Afficher plus de r√©sultats"
    }
}
