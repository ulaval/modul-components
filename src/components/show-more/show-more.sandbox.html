<div>
    <h3>m-show-more - With nbTotal > 0</h3>
    <m-show-more :nb-visible="data.length"
                 :nb-total="total"
                 :loading.sync="loading"
                 @click="fetchData">
    </m-show-more>

    <h3>m-show-more - With nbTotal = undefined</h3>
    <m-show-more :nb-visible="totalZero"
                 :nb-total="totalUndefined"
                 :loading.sync="loading"
                 @click="fetchData">
    </m-show-more>

    <h3>m-show-more - With nbTotal = 0</h3>
    <m-show-more :nb-visible="totalZero"
                 :nb-total="totalZero"
                 :loading.sync="loading"
                 @click="fetchData">
    </m-show-more>
</div>
