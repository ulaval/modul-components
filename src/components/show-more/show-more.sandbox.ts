import Vue, { PluginObject } from 'vue';
import Component from 'vue-class-component';
import { SHOW_MORE_NAME } from '../component-names';
import ShowMorePlugin from './show-more';
import WithRender from './show-more.sandbox.html';

@WithRender
@Component
export class MShowMoreSandbox extends Vue {
    data: number[] = [];
    total: number = 20;
    totalZero: number = 0;
    totalUndefined: undefined = undefined;
    loading: boolean = false;

    fetchData(): void {
        if (this.data.length < this.total) {
            setTimeout(() => {
                this.data.push(0, 1, 2, 3, 4, 5, 6, 7, 8, 9);
                this.loading = false;
            }, 1000);
        }
    }
}

const ShowMoreSandboxPlugin: PluginObject<any> = {
    install(v, options): void {
        v.use(ShowMorePlugin);
        v.component(`${SHOW_MORE_NAME}-sandbox`, MShowMoreSandbox);
    }
};

export default ShowMoreSandboxPlugin;
