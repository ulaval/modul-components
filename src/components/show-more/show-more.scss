@import 'commons';

.m-show-more {
    text-align: center;
    max-width: $m-min-width;

    &__status {
        margin-bottom: $m-spacing;
    }

    .m-button {
        margin-top: $m-spacing;
    }
}
