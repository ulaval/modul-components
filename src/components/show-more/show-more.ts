import { PluginObject } from 'vue';
import Component from 'vue-class-component';
import { Emit, Prop } from 'vue-property-decorator';
import { FormatMode } from '../../utils/i18n/i18n';
import { ModulVue } from '../../utils/vue/vue';
import { MButtonSkin } from '../button/button';
import { SHOW_MORE_NAME } from '../component-names';
import WithRender from './show-more.html?style=./show-more.scss';

@WithRender
@Component
export class MShowMore extends ModulVue {
    @Prop({ default: 0 })
    nbVisible: number;

    @Prop({ required: true })
    nbTotal: number;

    @Prop({ default: false })
    loading: boolean;

    get isVisible(): boolean {
        return this.nbTotal !== undefined && this.nbTotal > 0;
    }

    get status(): string {
        return this.$i18n.translate('m-show-more:status', { nbVisible: this.nbVisible, nbTotal: this.nbTotal }, undefined, undefined, undefined, FormatMode.Sprintf);
    }

    get visiblePourcentage(): number {
        return (this.nbVisible / this.nbTotal) * 100;
    }

    get buttonLabel(): string {
        return this.$i18n.translate('m-show-more:button-label');
    }

    get buttonSkin(): string {
        return MButtonSkin.Secondary;
    }

    get isButtonVisible(): boolean {
        return this.nbVisible < this.nbTotal;
    }

    @Emit('click')
    showMore(): void {
        this.$emit('update:loading', true);
    }
}

const ShowMorePlugin: PluginObject<any> = {
    install(v, options): void {
        v.prototype.$log.debug(SHOW_MORE_NAME, 'plugin.install');
        v.component(SHOW_MORE_NAME, MShowMore);
    }
};

export default ShowMorePlugin;
