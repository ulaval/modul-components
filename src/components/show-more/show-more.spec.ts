import { RefSelector, shallow, Wrapper } from '@vue/test-utils';
import { renderComponent } from '../../../tests/helpers/render';
import { FormatMode } from '../../utils/i18n/i18n';
import { MButtonSkin } from '../button/button';
import { SHOW_MORE_NAME } from '../component-names';
import { MShowMore } from './show-more';

let wrapper: Wrapper<MShowMore>;

const REF_COUNT_TOTAL: RefSelector = { ref: 'countTotal' };
const REF_PROGRESS: RefSelector = { ref: 'progress' };
const REF_BUTTON: RefSelector = { ref: 'button' };

const NB_VISIBLE: number = 10;
const NB_TOTAL: number = 50;

const initializeShallowWrapper: any = () => {
    wrapper = shallow(MShowMore);

    wrapper.vm.$i18n.translate = jest.fn((key: string) => key);
};

describe(SHOW_MORE_NAME, () => {
    describe(`Without nbTotal specified`, () => {
        it(`Should not render anything`, () => {
            initializeShallowWrapper();
            expect(renderComponent(wrapper.vm)).resolves.toMatchSnapshot();
        });
    });

    describe(`With nbTotal equals to 0`, () => {

        beforeEach(() => {
            initializeShallowWrapper();
            wrapper.setProps({ nbVisible: 0, nbTotal: -2 });
        });

        it(`Should not render anything`, () => {
            expect(renderComponent(wrapper.vm)).resolves.toMatchSnapshot();
        });

        it(`Should not be visible`, () => {
            expect(wrapper.vm.isVisible).toBeFalsy();
        });
    });

    describe(`With nbTotal inferior to 0`, () => {

        beforeEach(() => {
            initializeShallowWrapper();
            wrapper.setProps({ nbVisible: 0, nbTotal: -2 });
        });

        it(`Should not render anything`, () => {
            expect(renderComponent(wrapper.vm)).resolves.toMatchSnapshot();
        });

        it(`Should not be visible`, () => {
            expect(wrapper.vm.isVisible).toBeFalsy();
        });
    });

    describe(`Given nbVisible and nbTotal`, () => {
        beforeEach(() => {
            initializeShallowWrapper();
            wrapper.setProps({ nbVisible: NB_VISIBLE, nbTotal: NB_TOTAL });
        });

        it(`Should calculate pourcentage of visible elements`, () => {
            expect(wrapper.vm.visiblePourcentage).toBe(20);
        });

        it(`Should render a label with nbVisible and nbTotal`, () => {
            expect(wrapper.find(REF_COUNT_TOTAL).exists()).toBeTruthy();
            expect(wrapper.find(REF_COUNT_TOTAL).text()).toBe('m-show-more:status');
            expect(wrapper.vm.$i18n.translate).toHaveBeenCalledWith('m-show-more:status', { nbVisible: NB_VISIBLE, nbTotal: NB_TOTAL }, undefined, undefined, undefined, FormatMode.Sprintf);
        });

        it(`Should render a progress bar with value and border-radius props`, () => {
            expect(wrapper.find(REF_PROGRESS).exists()).toBeTruthy();
            expect(wrapper.find(REF_PROGRESS).attributes().value).toBe('20');
        });
    });

    describe(`Given nbVisible < nbTotal`, () => {
        beforeEach(() => {
            initializeShallowWrapper();
            wrapper.setProps({ nbVisible: NB_VISIBLE, nbTotal: NB_TOTAL });
        });

        it(`Should render correctly`, () => {
            expect(renderComponent(wrapper.vm)).resolves.toMatchSnapshot();
        });

        it(`Should be visible`, () => {
            expect(wrapper.vm.isVisible).toBeTruthy();
        });

        it(`Should render a button`, () => {
            expect(wrapper.find(REF_BUTTON).exists()).toBeTruthy();
            expect(wrapper.find(REF_BUTTON).text()).toBe('m-show-more:button-label');
            expect(wrapper.find(REF_BUTTON).attributes().skin).toBe(MButtonSkin.Secondary);
        });

        describe(`When button is clicked`, () => {
            it(`Should emit a "click" event`, () => {
                wrapper.find(REF_BUTTON).trigger('click');

                expect(wrapper.emitted('click')).toBeTruthy();
            });

            it(`Should emit a "update:loading" event with a value of true`, () => {
                wrapper.find(REF_BUTTON).trigger('click');

                expect(wrapper.emitted('update:loading')[0]).toEqual([true]);
            });
        });

        describe(`When loading`, () => {
            it(`Should put button in waiting mode`, () => {
                wrapper.setProps({ loading: true });

                expect(wrapper.find(REF_BUTTON).attributes().waiting).toBe('true');
            });
        });
    });

    describe(`Given nbVisible == nbTotal`, () => {
        beforeEach(() => {
            initializeShallowWrapper();
            wrapper.setProps({ nbVisible: NB_TOTAL, nbTotal: NB_TOTAL });
        });

        it(`Should render correctly`, () => {
            expect(renderComponent(wrapper.vm)).resolves.toMatchSnapshot();
        });

        it(`Should not render a button`, () => {
            expect(wrapper.find(REF_BUTTON).exists()).toBeFalsy();
        });
    });
});
