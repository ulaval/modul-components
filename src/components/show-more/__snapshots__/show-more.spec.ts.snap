// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`m-show-more Given nbVisible < nbTotal Should render correctly 1`] = `
<div class="m-show-more">
  <div class="m-show-more__status">m-show-more:status</div>
  <m-progress value="20"></m-progress>
  <m-button skin="secondary">m-show-more:button-label</m-button>
</div>
`;

exports[`m-show-more Given nbVisible == nbTotal Should render correctly 1`] = `
<div class="m-show-more">
  <div class="m-show-more__status">m-show-more:status</div>
</div>
`;

exports[`m-show-more With nbTotal equals to 0 Should not render anything 1`] = `""`;

exports[`m-show-more With nbTotal inferior to 0 Should not render anything 1`] = `""`;

exports[`m-show-more Without nbTotal specified Should not render anything 1`] = `""`;
